### PR TITLE
Hotfix - webpack & bundle optimizations

### PR DESCRIPTION
**Pull request by ArcanisCz**
Hi again! This PR is about improving build speed a bit, then about optimizing bundle. I will describe it in 2 sections:

1. Speedup (used https://github.com/stephencookdev/speed-measure-webpack-plugin)
2. Budle optimization (used https://github.com/webpack-contrib/webpack-bundle-analyzer)

I have tested it only in localhost development mode. Test it in some pseudo-production test/staging enviroment would be neccessary to be sure.

### Speedup
(all times are measured on my personal computer, absolute numbers might vary)
One of the most time-consuming operations was ImageminPlugin (around 26s). Fortunately, plugin has `cacheFolder` option, which allowed to speedup second run to 8s.

Another big consumer of time was css loader stack, which i didnt solve. Only discovered that node-sass probably has some cache (https://stackoverflow.com/questions/14934800/why-does-sass-cache-folder-get-created). Dont know if it is in effect or not.

A small speedup was a refactor of UglifyJs plugin according to webpack 4 docu (https://webpack.js.org/configuration/optimization/#optimization-minimizer). I suspected that it ran twice - once natively, once by adding as a plugin.

Big speedup was gained by enabling UglifyJs cache.

Small speedup was gained by moving in-app vendor code into vendors.

**Result: dist (`npm run dist`, without vendors, without vendorCheckSum.hs) build from around 1m 30s to around 30s.** (second run, witch caches)

### Bundle
When tried to check bundle composition, this horrible image was thrown upon me with number of issues
proton-bundle-old

Theese 2 chunks should only contain application code, not libraries (since you are building them separately). But they contain almost half of their size of vendor code! So i tried to do 2 things:

1. create a common chunk for app and appLazy, which will contain all the common vendor code for both of them. Theese will solve duplication of for example lodash. In best case scenario, eliminate this chunk.
2. Investigate cause and remove all possible code from app and appLazy.

It resulted in much nicer picture
proton-bundle-new

Most of the issues were solved by moving respective libraries into vendors. Since you build vendor separately and outside of webpack module system, you must take care about any and every `require` which is library. Webpack will include it even if it is in vendors (ical.js).

Issues which remained in vendor-app chunk (solved, it would probably eliminate vendor-app chunk):
- library `juice` - cannot be moved into vendor, because it has no dist bundle. Can be used only via require. But has its own bundled dependencies (rly?) like cheerio, which bloats bundle horribly. I would consider some alternative, or write code by yourselves.
- lodash - i would consider bundling whole lodash into vendors, and drop that selective imports in code. Application uses quite a range of features.

Btw, this would be a bundle without `juice`
proton-bundle-juice

**result:**
(`npm run dist`)

  | old bundle (kb) | new bundle (kb)
-- | -- | --
app.js (minimized) | 2 774 | 1 173
appLazy.js (minimized) | 1 553 | 1 060
vendor-app.js (minimized) | 0 | 374
vendor.js (unminimized) | 3 934 | 4 027
vendorLazy.js (unminimized) | 1 304 | 1 038
vendorLazy2.js (unminimized) | 399 | 399
sum  | 9 964 | 8 071

### CODE REVIEW

File: env/conf.build.js [hunk 1]
```diff
@@ -47,7 +50,6 @@ module.exports = {
             'jquery-timepicker/jquery.timepicker.js',
             'squire-rte/build/squire-raw.js',
             'dropzone/dist/dropzone.js',
-            'ical.js/build/ical.js',
```
Comment (ArcanisCz) [inline env/conf.build.js:50]: Duplicate from js

File: package.json [hunk 2]
```diff
@@ -176,7 +176,6 @@
     "i18n:build": "npm run i18n:extract && npm run i18n:update && npm run i18n:compile",
     "posti18n:build": "./tasks/createI18nCache.js && ./tasks/commitI18n.sh",
     "pretty": "prettier --write $(find src/app -type f -name '*.js')",
-    "precommit": "./tasks/metrics.sh",
```
Comment (ArcanisCz) [inline package.json:179]: sorry, could not commit with this on my windows :(

File: src/app/app.js [hunk 3]
```diff
@@ -24,8 +23,6 @@ import CONFIG from './config';
 import routes from './routes';
 import '../sass/app.scss';
 
-window.moment = timezone;
```
Comment (ArcanisCz) [inline src/app/app.js:27]: i am not sure if this is equivalent. But importing moment-timezone will cause to bundle it with app. It seems from docu, it works with plain without reassigning moment https://github.com/moment/moment-timezone

File: src/app/attachments/services/AttachmentEvent.js [hunk 4]
```diff
@@ -1,5 +1,3 @@
-import ICAL from 'ical.js';
```
Comment (ArcanisCz) [inline src/app/attachments/services/AttachmentEvent.js:1]: global variable from vendors. this import duplicated ical.

File: src/app/bugReport/services/bugReportApi.js [hunk 5]
```diff
@@ -1,5 +1,4 @@
 import _ from 'lodash';
-import Raven from 'raven-js';
```
Comment (ArcanisCz) [inline src/app/bugReport/services/bugReportApi.js:2]: global from vendors

File: webpack.tasks/js.loader.js [hunk 6]
```diff
@@ -17,17 +17,8 @@ const jsLoader = () => {
...
-                    'env',
-                    {
-                        targets: {
-                            browsers: ['ie 11']
-                        }
-                    }
-                ]
-            ],
-            plugins: [require('babel-plugin-transform-object-rest-spread'), require('babel-plugin-transform-runtime')],
+            presets: ['env'],
```
Comment (ArcanisCz) [inline webpack.tasks/js.loader.js:20]: THeese 2 lines eliminates babel-polyfill (core-js) from app-vendor. Its uncessary to include only polyfills for IE11, if you are including whole polyfill in vendors anyway. If needed only some polyfills, you should remove babel-polyfill in vendors, and leave this here.

File: webpack.config.js [hunk 7]
```diff
@@ -27,6 +27,19 @@ module.exports = {
     watchOptions: {
         ignored: [/node_modules/, 'i18n/*.json', /\*\.(gif|jpeg|jpg|ico|png)/]
     },
+    optimization: {
+        minimizer: [require('./webpack.tasks/uglify.plugin')],
```
Comment (ArcanisCz) [inline webpack.config.js:31]: webpack will minimize only in production mode

File: webpack.tasks/uglify.plugin.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+const minifyJS = () => {
+    return {
+        mangle: true,
+        compress: true
+    };
+};
+
+const uglifyPlugin = new UglifyJSPlugin({
+    exclude: /\/node_modules/,
+    cache: true,
```
Comment (ArcanisCz) [inline webpack.tasks/uglify.plugin.js:12]: its false by default https://github.com/webpack-contrib/uglifyjs-webpack-plugin#cache

File: webpack.config.js [hunk 9]
```diff
@@ -27,6 +27,18 @@ module.exports = {
...
+        splitChunks: {
```
Comment (ArcanisCz) [inline webpack.config.js:32]: https://webpack.js.org/plugins/split-chunks-plugin/#split-chunks-example-2

**Comment by dhoko** [comment]
:+1: Nice work, thx.
We will try to add this inside the project soon, it will be better for everyone. I will take a look this week. 
We have some things to improve too, such as loading templates/modules instead of everything inside the main app.js etc.

2 774 | 1 173
-- | --

^ this, :100: 


9 964 | 8 071
-- | --

O_o the diff is huge.

Yes juice is not a good lib for a browser, I saw cheerio inside too but we didn't didn't found something else. Looking at the code I think we can manage to create a plugin for DOMPurify from it, it would be easier to use for us, and it won't require cheerio.

Anyway, thx for this :+1: (_sorry for the delay, quite busy this last weeks_ ;) )

File: webpack.tasks/uglify.plugin.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
```
Comment (Pegase745) [inline webpack.tasks/uglify.plugin.js:1]: How about using [terser-webpack-plugin](https://github.com/webpack-contrib/terser-webpack-plugin) as webpack@5 will be using the terser plugin by default?

Comment (dhoko) [reply webpack.tasks/uglify.plugin.js:1]: Yup, it's inside the todolist ;)

**Pull request closed without merging**